src/primitive.rs

```diff
@@ -0,0 +1,213 @@
+// Copyright (c) 2018 King's College London created by the Software Development
+// Team <http://soft-dev.org/>
+//
+// The Universal Permissive License (UPL), Version 1.0
+//
+// Subject to the condition set forth below, permission is hereby granted to any
+// person obtaining a copy of this software, associated documentation and/or
+// data (collectively the "Software"), free of charge and under any and all
+// copyright rights in the Software, and any and all patent rights owned or
+// freely licensable by each licensor hereunder covering either (i) the
+// unmodified Software as contributed to or provided by such licensor, or (ii)
+// the Larger Works (as defined below), to deal in both
+//
+// (a) the Software, and (b) any piece of software and/or hardware listed in the
+// lrgrwrks.txt file if one is included with the Software (each a "Larger Work"
+// to which the Software is contributed by such licensors),
+//
+// without restriction, including without limitation the rights to copy, create
+// derivative works of, display, perform, and distribute the Software and make,
+// use, sell, offer for sale, import, export, have made, and have sold the
+// Software and the Larger Work(s), and to sublicense the foregoing rights on
+// either these or other terms.
+//
+// This license is subject to the following condition: The above copyright
+// notice and either this complete permission notice or at a minimum a reference
+// to the UPL must be included in all copies or substantial portions of the
+// Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+use byteorder::{ByteOrder, LittleEndian};
+
+use rustc_const_math::ConstInt;
+use rustc::mir::{Operand, Constant, Literal};
+use rustc::middle::const_val::ConstVal;
+use rustc::ty::{Ty, TypeVariants};
+use syntax::ast::{IntTy, UintTy};
+
+use interp::TypedVal;
+use machine::Machine;
+
+/// An operand's value can always be determined without needing to temporarily
+/// save values to a stack. This is because of the recursive nature of the MIR.
+/// An operand will only ever return a primitive value or a pointer.
+pub fn eval_operand<'a,'tcx>(operand: &Operand<'tcx>,
+                             vm: &mut Machine<'a, 'tcx>)
+                             -> TypedVal<'tcx> {
+    let ty = operand.ty(vm.cur_frame().mir, vm.tcx);
+    match *operand {
```

Comment:
It's probably worth remember that, in general, `match` statements will get converted to code that's basically "if A { ... } elif B { ... }` and so on. AFAIK these will end up in the order specified in the user's file. So you probably want to put the least common case (e.g. the `unreachable` arm) last.

Reply:
Sure I understand. But I'm a bit confused as to the relevance of this comment against this specific section of code. If I understand correctly, are you saying that the match arms on operators should match the order of precedence specified by rust?

Reply:
No, I'm saying you probably want to put the least likely to fire case at the end.

Reply:
Okay, I'll bear this in mind in future. In the particular case of operands, I'm not sure which would be the most likely case, so I'll leave this in the same order as is written in the rust compiler.